### PR TITLE
xsdgen: better type deduplication

### PR DESCRIPTION
**Pull request by SoMuchForSubtlety**
There is no need to deduplicate types that are equivalent

### CODE REVIEW

File: xsd/xsd.go [hunk 1]
```diff
@@ -38,6 +39,107 @@ type Type interface {
 	isType()
 }
 
+func TypesEqual(a, b Type) bool {
+	switch a := a.(type) {
+	case *SimpleType:
+		switch b := b.(type) {
+		case *SimpleType:
```
Comment (erendogan51) [inline xsd/xsd.go:46]: isnt this redundant?

File: xsd/xsd.go [hunk 2]
```diff
@@ -38,6 +39,107 @@ type Type interface {
...
+		a.Length != b.Length || a.MinLength != b.MinLength || a.MaxLength != b.MaxLength ||
+		a.MinDate != b.MinDate || a.MaxDate != b.MaxDate ||
+		a.Precision != b.Precision ||
+		a.TotalDigits != b.TotalDigits {
+		return false
+	}
+
+	if (a.Pattern != nil && b.Pattern == nil) ||
+		(a.Pattern == nil && b.Pattern != nil) ||
+		(a.Pattern != nil && a.Pattern.String() != b.Pattern.String()) {
```
Comment (erendogan51) [inline xsd/xsd.go:114]:
```suggestion
		(a.Pattern != nil && b.Pattern != nil &&  a.Pattern.String()  != b.Pattern.String()) {
```

File: xsd/xsd.go [hunk 3]
```diff
@@ -38,6 +39,107 @@ type Type interface {
 	isType()
 }
 
+func TypesEqual(a, b Type) bool {
```
Comment (erendogan51) [inline xsd/xsd.go:42]:
```suggestion
func typesEqual(a, b Type) bool {
```

Comment (SoMuchForSubtlety) [reply xsd/xsd.go:42]: This needs to be exported, we call it from a different package

File: xsd/xsd.go (hunk 2 above)
Comment (SoMuchForSubtlety) [reply xsd/xsd.go:114]: What if only one of them has a pattern?

File: xsd/xsd.go (hunk 1 above)
Comment (SoMuchForSubtlety) [reply xsd/xsd.go:46]: no, we need to reassign `b` in order to be able to use it as the matched type

File: xsd/xsd.go (hunk 2 above)
Comment (erendogan51) [reply xsd/xsd.go:114]: but calling `.Pattern.String()` on a nil `Pattern` would panic right?

Comment (SoMuchForSubtlety) [reply xsd/xsd.go:114]: correct

Comment (erendogan51) [reply xsd/xsd.go:114]: yeah then you have to check if both a and b patterns a not nil

Comment (SoMuchForSubtlety) [reply xsd/xsd.go:114]: Oh, sorry, I misread your suggestion. Only one pattern being nil would be caught by the first two checks.

**Review by erendogan51 — APPROVED** [review]

**Pull request merged**